File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/CreateCategoryController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.dto.CategoryDTO
import net.animeta.backend.model.User
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.CategorySerializer
import net.animeta.backend.service.CategoryMutations
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class CreateCategoryController(
    private val categoryMutations: CategoryMutations,
    private val categorySerializer: CategorySerializer
) {
    data class Params(
        val name: String
    )
    data class Result(
        val category: CategoryDTO
    )

    @PostMapping("/v3/CreateCategory")
    @Transactional
    fun handle(@RequestBody params: Params, @CurrentUser currentUser: User): Result {
        val category = categoryMutations.create(currentUser, params.name)
        return Result(categorySerializer.serialize(category))
    }
}

File: backend-ts/src/controllers/create_category.controller.ts
import { Body, Controller, Post } from "@nestjs/common";
import { CategoryDTO } from 'shared/types';
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { CategorySerializer } from "src/serializers/category.serializer";
import { CategoryService } from "src/services/category.service";
import { Connection } from "typeorm";

type Params = {
  name: string;
}

type Result = {
  category: CategoryDTO;
}

@Controller('/api/v4/CreateCategory')
export class CreateCategoryController {
  constructor(
    private connection: Connection,
    private categoryService: CategoryService,
    private categorySerializer: CategorySerializer,
  ) {}

  @Post()
  async handle(
    @Body() params: Params,
    @CurrentUser() currentUser: User,
  ): Promise<Result> {
    return await this.connection.transaction(async em => {
      const category = await this.categoryService.createCategory(em, currentUser, params)
      return { category: this.categorySerializer.serialize(category) }
    })
  }
}
